### DISCUSSION TITLE
how to create a python package with a dll/so file ?

### DISCUSSION THREAD

**Discussion by hbadi** [pybind/pybind11#3394] [Q&A]
Hello,

I've been reading some stuffs last days about python module packaging,  it's quite a mess to understand something clearly since there's a lot of package management tools like setuptools.

I have my own cmake file to generate my dynamic library file. I can't see how to write a very basic setup.py in which i only specify the dynamic library file to use. I found many examples/tutorials in which Extension/pybind11 python modules are used through the explicit definition of source files, but in my case i want to compile the dynlib by my own. Is there any simple setup.py to do the job ?

For example under windows i tried to write these setup.py but the dll is not copied...i only get the .egg file

```
from setuptools import setup, Extension
setup(
    name='mymodule',
    version='1.0',
    description='A Python package C++ (PyBind11)',
    data_files=[('lib\\site-packages\\', ["mymodule.dll"])],
)
```
`python setup.py install`

The website [https://pybind11.readthedocs.io/en/stable/compiling.html](https://pybind11.readthedocs.io/en/stable/compiling.html) do explain a lot of things but only where there's a list of cpp files, which is not my case that's why i manage the compiling using cmake.
I can use `ext_modules = [Extension('pysparselizard', sources=[''])]` but source files are needed.

It's not very clear...

Thansk for your help.

**Comment by LivelyLiz** [comment]
Did you find a solution? I am trying to find out the same thing but also mainly find resources for doing this with .py files...

**Comment by matte86** [comment]
I have a similar setup where I build my project with conan and cmake. The project is structured in a way so that I have:

- A shared cpp library
- A shared library containing the bindings with Python (as output of the pybind11 binding build process)

The best process I managed to implement so far is to build everythin from conan and cmake and then use the cmake install routine to create a specific python_package folder containing all my targets, placed in the proper directories. Inside my setup.py file I then have something like:

```
...
package_dir={"": "."},
package_data={'': ['*.so', "*.dylib"]},
...
```

I am not an expert of python binary packages and I am still playing around a lot. The process I implemented so far allows me to properly package shared libs into my python package but I am not entirely sure this approach is correct when it comes to distributing my package.

**Comment by GabrielJMS** [comment]
Hello @hbadi, did you find a solution for this?

**Reply by Deepak-98M** [reply]
waiting for a solution here as well, apparently the visual studio build tools do not help as they show error such as in my case 'sys/time.h' is not available, as it is especially for windows, i have a library that needs an additinal extension library which is in .c format and then converted to .so and not recognized.

**Comment by PhilipNelson5** [comment]
I've had success in the past, when working with SWIG, creating a python package from the compiled *.pyd files using a pyproject.toml file.

_file structure_
```
build/
    python/
        moduleName/
            moduleName.pyd
        LICENSE
        pyproject.toml
```

_pyproject.toml_
```toml
[project]
name = "moduleName"
license-files = ["LICENSE"]
version = "@PROJECT_VERSION@"
dependencies = []

[tool.setuptools]
packages = ["moduleName"]

[tool.setuptools.package-data]
moduleName = ["*.pyd"]

[build-system]
requires = ["setuptools >= 77.0"]
build-backend = "setuptools.build_meta
```

However, when I use this same approach with pybind11, I can import the module but none of the functions exist...

When I add the build directory to the python system path, `sys.path.append("build/python/moduleName")` it works like a charm.

When I `pip install build/python`, I get `AttributeError: module 'moduleName' has no attribute 'foo'`

**Comment by anthonygclark** [comment]
I was able to do this improperly (probably).

Layout:
```
python-mypackage/
   setup.py
   mypackage/
      __init__.py
      mypackage.so
```

The `mypackage.so` is built via pybind11 using the suggested manual build steps (by invoking g++ manually, etc).

The setup.py looks like the following. At this point, I was flailing and trying all sorts of setup() parameters. Advice on Setuptools and packaging extensions is terse. This should probably move to pyproject.toml to avoid the constant blabbing about setuptools ;) :
```py
from setuptools import setup

setup(
        name = 'mypackage',
        version  = '1.0.0',
        author = '',
        packages   = ['mypackage'],
        include_package_data = True,
        package_data = { 'mypackage': ['mypackage/*.so'] },
        zip_safe  = False
)
```

`__init__.py` has this:
```py
import sys
import importlib.resources
sys.path.append(str(importlib.resources.files(__package__)))
del importlib
del sys

# you might have to play around with this to get your 
# desired effect.
import mymodule.mymodule as mymodule
```

I tested it with a venv:
```sh
$ python -m venv venv
$ source venv/bin/activate
(venv) $ pip install python-mymodule
(venv) $ python
>>> import mymodule 
```

1. I'm not sure if adding package_data or package_dir in the setup() call is the best way to do this. Also, I don't know if I just got lucky with this layout and it only works for a single module - etc.
3. good luck - and, sorry if this is very oldschool or using deprecated components.
